wandb/sdk/internal/sender.py

```diff
@@ -890,16 +948,117 @@ def send_stats(self, record: "Record") -> None:
         self._fs.push(filenames.EVENTS_FNAME, json.dumps(row))
         # TODO(jhr): check fs.push results?
 
+    def _output_raw_finish(self) -> None:
+        for stream, output_raw in self._output_raw_streams.items():
+            output_raw._stopped.set()
+
+            # shut down threads
+            output_raw._writer_thr.join(timeout=5)
+            if output_raw._writer_thr.is_alive():
+                logger.info("processing output...")
+                output_raw._writer_thr.join()
+            output_raw._reader_thr.join()
+
+            # flush output buffers and files
+            self._output_raw_flush(stream)
+        self._output_raw_streams = {}
+        if self._output_raw_file:
+            self._output_raw_file.close()
+            self._output_raw_file = None
+
+    def _output_writer_thread(self, stream: "StreamLiterals") -> None:
```

Comment:
This code is modeled closely after the existing `WRAP` implementation.  small cleanups, type annotations and fixes were made but because of the difficulty of testing all these conditions it was safest to keep the functionality as close to the same as possible.    existing wrap implementation that was the inspiration is in wandb_run and lib/redirect.py

Reply:
could you leave some documentation in the code now that you understand this logic better

Reply:
Instead it is in this working doc:
https://www.notion.so/wandbai/wandb-console-4d843f0de0404d88a47eb04e6bfb926a